arch/BPF/BPFDisassembler.c

```diff
@@ -407,52 +401,482 @@ static bool getInstruction(cs_struct *ud, MCInst *MI, bpf_internal *bpf)
+	}
+
+	return BPF_INS_INVALID;
+}
+#undef SPEC_CASE
+#undef CASE
+#undef BPF_CALLX
+
+#ifndef CAPSTONE_DIET
+
+inline static void update_regs_read(cs_detail *detail, const bpf_reg r)
+{
+	detail->regs_read[detail->regs_read_count] = r;
+	detail->regs_read_count++;
+}
+
+inline static void update_regs_write(cs_detail *detail, const bpf_reg r)
+{
+	detail->regs_write[detail->regs_write_count] = r;
+	detail->regs_write_count++;
+}
+
+static void update_regs_access(const MCInst *MI, cs_detail *detail,
+			       bpf_insn insn_id, unsigned int opcode)
+{
+	if (insn_id == BPF_INS_INVALID)
+		return;
+	/*
+	 * In eBPF mode, only these instructions have implicit registers access:
+	 * - legacy ld{w,h,b,dw} * // w: r0
+	 * - exit // r: r0
+	 */
+	if (EBPF_MODE(MI->csh->mode)) {
+		switch (insn_id) {
+		default:
+			break;
+		case BPF_INS_LDABSW:
+		case BPF_INS_LDABSH:
+		case BPF_INS_LDABSB:
+		case BPF_INS_LDINDW:
+		case BPF_INS_LDINDH:
+		case BPF_INS_LDINDB:
+		case BPF_INS_LDDW:
+			if (BPF_MODE(opcode) == BPF_MODE_ABS ||
+			    BPF_MODE(opcode) == BPF_MODE_IND)
+				update_regs_write(detail, BPF_REG_R0);
+			break;
+		case BPF_INS_EXIT:
+			update_regs_read(detail, BPF_REG_R0);
+			break;
+		}
+		return;
+	}
+
+	/* cBPF mode */
+	switch (BPF_CLASS(opcode)) {
+	default:
+		break;
+	case BPF_CLASS_LD:
+		update_regs_write(detail, BPF_REG_A);
+		break;
+	case BPF_CLASS_LDX:
+		update_regs_write(detail, BPF_REG_X);
+		break;
+	case BPF_CLASS_ST:
+		update_regs_read(detail, BPF_REG_A);
+		break;
+	case BPF_CLASS_STX:
+		update_regs_read(detail, BPF_REG_X);
+		break;
+	case BPF_CLASS_ALU:
+		update_regs_read(detail, BPF_REG_A);
+		update_regs_write(detail, BPF_REG_A);
+		break;
+	case BPF_CLASS_JMP:
+		if (insn_id != BPF_INS_JA) // except the unconditional jump
+			update_regs_read(detail, BPF_REG_A);
+		break;
+	/* case BPF_CLASS_RET: */
 	case BPF_CLASS_MISC:
-	/* case BPF_CLASS_ALU64: */
-		if (EBPF_MODE(ud))
-			return decodeALU(ud, MI, bpf);
+		if (insn_id == BPF_INS_TAX) {
+			update_regs_read(detail, BPF_REG_A);
+			update_regs_write(detail, BPF_REG_X);
+		} else {
+			update_regs_read(detail, BPF_REG_X);
+			update_regs_write(detail, BPF_REG_A);
+		}
+		break;
+	}
+}
+#endif
+
+#ifndef CAPSTONE_DIET
+static inline void push_group(cs_detail *detail, const bpf_insn_group group)
+{
+	if (!detail)
+		return;
+
+	detail->groups[detail->groups_count] = group;
+	detail->groups_count++;
+}
+#else
+#define push_group(detail, group)
+#endif
+
+static bool setFinalOpcode(MCInst *MI, const bpf_internal *bpf)
+{
+	bpf_insn id = BPF_INS_INVALID;
+#ifndef CAPSTONE_DIET
+	cs_detail *detail;
+
+	detail = get_detail(MI);
+#endif
+
+	const uint16_t opcode = bpf->op;
+	switch (BPF_CLASS(opcode)) {
+	default: // will never happen
+		break;
+	case BPF_CLASS_LD:
+	case BPF_CLASS_LDX:
+		if (EBPF_MODE(MI->csh->mode))
+			id = op2insn_ld_ebpf(opcode);
 		else
-			return decodeMISC(ud, MI, bpf);
+			id = op2insn_ld_cbpf(opcode);
+		push_group(detail, BPF_GRP_LOAD);
```

Comment:
```suggestion
		add_group(MI, BPF_GRP_LOAD);
```

Also below.

Reply:
Fixed the `add_group`.
Why should we remove line 435?